### ISSUE TITLE
r5r Not recognizing Java Version

### ISSUE THREAD

**Issue by willj98** [ipeaGIT/r5r#393]
Hello,

I apologize if this is an overly simple question, but I have not been able to figure it out on my own after lots of trying, and I am not familiar with java.

My issue is that no matter what version Java jdk I download (21, 22), I get this error from r5r when trying to use the `setup_r5 `function : 
Error in setup_r5(data_path, verbose = TRUE) : 
  This package requires the Java SE Development Kit 21.
Please update your Java installation. The jdk 21 can be downloaded from either:
  - openjdk: https://jdk.java.net/java-se-ri/21
  - oracle: https://docs.oracle.com/en/java/javase/21/install/index.html

I am on an Intel 64 macbook. I have reinstalled R and and ensured that my home environment is set correctly. To be sure, I used the helpful java quick installation provided by [e-kotov](https://github.com/e-kotov) : 
```
devtools::install_github("e-kotov/rJavaEnv")
rJavaEnv::java_quick_install(version = 21)
print(Sys.getenv("JAVA_HOME"))
```
Any help would be greatly appreciated! thank you in advance!!

**Comment by e-kotov** [comment]
Hi, @willj98 , could you please run in R console and show the output of:

```r
system("java --version")
system("which java")
``` 

and also to the same in macOS's Terminal.app?
```bash
java --version
which java
```

**Comment by e-kotov** [comment]
You may be affected by https://github.com/e-kotov/rJavaEnv/issues/8 , in case there was no Java installed on your computer before you installed R and `{r5r}` (which also triggers the install of `{rJava}`).

So if you could also show if:
```r
library(rJava)
```

loads, or if it gives you the error like that:
```
#> Error: package or namespace load failed for ‘rJava’:
#>  .onLoad failed in loadNamespace() for 'rJava', details:
#>   call: dyn.load(file, DLLpath = DLLpath, ...)
#>   error: unable to load shared object '/usr/local/lib/R/site-library/rJava/libs/rJava.so':
#>   libjvm.so: cannot open shared object file: No such file or directory
```

**Comment by willj98** [comment]
Thank you for your prompt reply. 
Running
 ```
system("java --version")
system("which java")
```
In the R terminal gives this output: 
> system("java --version")
openjdk 21.0.3 2024-04-16 LTS
OpenJDK Runtime Environment Corretto-21.0.3.9.1 (build 21.0.3+9-LTS)
OpenJDK 64-Bit Server VM Corretto-21.0.3.9.1 (build 21.0.3+9-LTS, mixed mode, sharing)
> system("which java")
/Users/will/Library/Caches/org.R-project.R/R/rJavaEnv/installed/macos/x64/21/bin/java

Running the code you suggested in the mac terminal gives me this output: 
java --version
which java
openjdk 11.0.22 2024-01-16
OpenJDK Runtime Environment Temurin-11.0.22+7 (build 11.0.22+7)
OpenJDK 64-Bit Server VM Temurin-11.0.22+7 (build 11.0.22+7, mixed mode)
/Users/will/JavaSE11/jdk-11.0.0.1/bin/java

I notice that there is a discrepancy between the two versions!

trying to load `rJava` in R using `library(rJava)` works correctly.

**Comment by e-kotov** [comment]
Than you should not be affected by the bug I referenced above.

The discrepancy between the versions is ok. You have Java 11 installed system wide (and that prevents the problem with `{rJava}` that I referenced earlier, where it does not load at all). And with my package you have installed the correct Java 21 that is picked up by `R` in your project directory.

Everything seems to be set up and working correctly on your system... Could you please also show the output of:

```r
rJavaEnv::java_check_version_cmd()
rJavaEnv::java_check_version_rjava()
```

**Comment by willj98** [comment]
Sure, here it is:
**> rJavaEnv::java_check_version_cmd()**
JAVA_HOME: /Users/will/Library/Caches/org.R-project.R/R/rJavaEnv/installed/macos/x64/21
Java path: /Users/will/Library/Caches/org.R-project.R/R/rJavaEnv/installed/macos/x64/21/bin/java
Java version: "openjdk version \"21.0.3\" 2024-04-16 LTS OpenJDK Runtime Environment Corretto-21.0.3.9.1 (build 21.0.3+9-LTS) OpenJDK
64-Bit Server VM Corretto-21.0.3.9.1 (build 21.0.3+9-LTS, mixed mode, sharing)"
[1] TRUE
**> rJavaEnv::java_check_version_rjava()**
Using current session's JAVA_HOME: /Users/will/Library/Caches/org.R-project.R/R/rJavaEnv/installed/macos/x64/21
With the user-specified JAVA_HOME rJava and other rJava/Java-based packages will use Java version: "21.0.3"
[1] TRUE

**Comment by e-kotov** [comment]
That is weird. It should be working. The `setup_r5()` should be working...

Here's what `setup_r5()` does internally:
```r
rJava::.jinit()
    ver <- rJava::.jcall("java.lang.System", "S", "getProperty", 
        "java.version")
    ver <- as.numeric(gsub("\\..*", "", ver))
    if (ver != 21) {
        stop("This package requires the Java SE Development Kit 21.\n", 
            "Please update your Java installation. ", "The jdk 21 can be downloaded from either:\n", 
            "  - openjdk: https://jdk.java.net/java-se-ri/21\n", 
            "  - oracle: https://docs.oracle.com/en/java/javase/21/install/index.html")
    }
```

This is the part of the code that leads to the message you got in the original post:
```
Error in setup_r5(data_path, verbose = TRUE) :
This package requires the Java SE Development Kit 21.
Please update your Java installation. The jdk 21 can be downloaded from either:

openjdk: https://jdk.java.net/java-se-ri/21
oracle: https://docs.oracle.com/en/java/javase/21/install/index.html
```


So does the same happen if you run this block of code manually?

Meanwhile, I will link to one possible solution in about 5-10 minutes.

**Comment by willj98** [comment]
Yes, running this block of code 
```
rJava::.jinit()
    ver <- rJava::.jcall("java.lang.System", "S", "getProperty", 
        "java.version")
    ver <- as.numeric(gsub("\\..*", "", ver))
    if (ver != 21) {
        stop("This package requires the Java SE Development Kit 21.\n", 
            "Please update your Java installation. ", "The jdk 21 can be downloaded from either:\n", 
            "  - openjdk: https://jdk.java.net/java-se-ri/21\n", 
            "  - oracle: https://docs.oracle.com/en/java/javase/21/install/index.html")
    }
```
returned a character object `ver` with value `"21.0.3"`

Thank you very much for your help, I really appreciate it!

**Comment by e-kotov** [comment]
I have no idea what's going on. The most I can suggest at the moment is to try installing

```r
remotes::install_github("e-kotov/r5r@no-java-check", subdir = "r-package", force = T)
```

This is a version of `r5r` that I have just made for you, that does not check the Java version and does not stop with error. Perhaps this will help.

**Comment by willj98** [comment]
Thank you! that allowed me to create the `r5r_core`. when I try to run the following code: 
```
> expanded_travel_time_matrix(
+   r5r_core,
+   origins,
+   destinations,
+   mode = "BUS",
+   mode_egress = "WALK",
+   departure_datetime = Sys.time(),
+   time_window = 10L,
+   breakdown = FALSE,
+   max_walk_time = Inf,
+   max_bike_time = Inf,
+   max_car_time = Inf,
+   max_trip_duration = 120L,
+   walk_speed = 3.6,
+   bike_speed = 12,
+   max_rides = 3,
+   max_lts = 2,
+   draws_per_minute = 5L,
+   n_threads = Inf,
+   verbose = TRUE,
+   progress = FALSE,
+   output_dir = "E:\\REU_work\\HealthStudy\\CurrentTransitTimes"
+ )
```
I get this error: 

```
Error in .jcall("RJavaTools", "Z", "hasField", .jcast(x, "java/lang/Object"),  : 
  java.lang.NullPointerException: Cannot invoke "Object.getClass()" because "<parameter1>" is null
```
I also got this error on my windows desktop yesterday, so hopefully it is something I am doing wrong, and an easier fix. I am not very knowledgeable yet concerning r5r., I apologize.

I believe I have confirmed that my GTFS and OSM .pbf data is correct.

**Comment by e-kotov** [comment]
@rafapereirabr has anyone else in the workshop experienced something like that? I'm lost, why `setup_r5()` could be failing the Java version check if `rJavaEnv::java_check_version_rjava()` says it's fine, as basically we are using the same approach, only `rJavaEnv::java_check_version_rjava()` spawns a separate R process to prevent initialising Java in the current R session.



@willj98 just in case (I'm sure you have done it multiple times already), also try to reload the session after:
```r
rJavaEnv::java_quick_install(version = 21)
rJavaEnv::java_check_version_rjava()
```

and strictly before:
```r
options(java.parameters = "-Xmx2G")
library(r5r)
r5r_core <- r5r::setup_r5(data_path, 
                          verbose = TRUE)
```

**Comment by willj98** [comment]
I restarted the session and ran those lines of code. this line 
`> rJavaEnv::java_check_version_rjava()` gives the output:
```
Using current session's JAVA_HOME:
/Users/will/Library/Caches/org.R-project.R/R/rJavaEnv/installed/macos/x64/21
With the user-specified JAVA_HOME rJava and other rJava/Java-based packages will use Java version:
"21.0.3"
[1] TRUE
```
I created the `r5_core` again and tried to use it in the `expanded_travel_time_matrix` and got this error again: 
```
Error in .jcall("RJavaTools", "Z", "hasField", .jcast(x, "java/lang/Object"),  : 
  java.lang.NullPointerException: Cannot invoke "Object.getClass()" because "<parameter1>" is null
```

**Comment by dhersz** [comment]
From what I understood from @willj98's last comment, he was able to use `setup_r5()` without an error, right? So that seems to suggest that now Java is installed correctly. To be sure, try running `setup_r5()` with `overwrite = TRUE`. 

The `java.lang.NullPointerException` error when trying to calculate a travel time matrix is probably related to other issues, not the Java installation.

**Comment by e-kotov** [comment]
@dhersz I think it is still related to Java version not being picked up correctly. I thought the problem was with the Java version checking code in `setup_r5()`, so I just disabled it. So now the version check is skipped, but that does not change the fact that wrong Java path is picked up. So even though my version of `setup_r5()` skips the Java check, it breaks when this incorrect Java is initialised. I am testing here: https://github.com/e-kotov/rJavaEnv/issues/8 and here: https://github.com/e-kotov/r5r-containerized

**Comment by e-kotov** [comment]
@willj98 ok, problem seems to be solved but will require some more steps from you (see option 2 below), or you can work online (option 1), but will be limited to 1GB memory.

**Option 1 (remote)**
You can run a temporary container with r5r in a web browser. Go here https://github.com/e-kotov/r5r-containerized , click the `launch binder` button in the repo and wait a bit. You should get an RStudio session where r5r works.

**Option 2 (your computer)**
This will involve some more manual steps:
1. Close all RStudio/r windows.
2. In Terminal.app:
```bash
export JAVA_HOME=/Users/will/Library/Caches/org.R-project.R/R/rJavaEnv/installed/macos/x64/21/
sudo R CMD javareconf
```

3. Run RStudio and `{r5r}` should now work.

At least this is from my testing in the repo linked above.

**Comment by willj98** [comment]
Running this code: 
```
export JAVA_HOME=/Users/will/Library/Caches/org.R-project.R/R/rJavaEnv/installed/macos/x64/21/
sudo R CMD javareconf
```
gives this output: 
```
Java interpreter : /Users/will/JavaSE11/jdk-11.0.0.1/bin/java
/Library/Frameworks/R.framework/Resources/bin/javareconf: line 106: /Users/will/JavaSE11/jdk-11.0.0.1/bin/java: cannot execute binary file

*** Java interpreter doesn't work properly.
```
I have no idea what is going on :(

**Comment by Audit-uCanB** [comment]
Is you program producing a network.dat with setup_r5() ?

**Comment by willj98** [comment]
@Audit-uCanB 
`setup_r5` creates a` osm.pbf.mapdb `and a `osm.pbf.mapdb.p` file. When I attempt to use `r5_core` in `expanded_travel_time_matrix` it gives me this error: 
```
Error in .jcall("RJavaTools", "Z", "hasField", .jcast(x, "java/lang/Object"),  : 
  java.lang.NullPointerException: Cannot invoke "Object.getClass()" because "<parameter1>" is null
```
and when I simply run `r5_core` in the terminal I get something like `rJava<Null>` (i am not at my computer right now so I am not sure if it is exactly that.)

**Comment by willj98** [comment]
@Audit-uCanB

as for a file with the actual extension network.dat , nothing is being created

**Comment by Audit-uCanB** [comment]
I've run into this error before with the parameter 1, I'm still sorting out whether there is a bigger bug with r5r 2.0 because it was just updated two months ago, now requiring Java 21. However, I got it to go away by doing the following two steps:

First, make sure your GTFS file ends with "GTFS", I believe it needs to be at the end of the name
Second, make sure you increase the ram allocation to Java at the very beginning of your code. If RStudio is open, end your session and open again in order to run it appropriately (it's picky and needs to be done before loading r5r package)

`options(java.parameters` = "-Xmx5G")   

You can replace the 5 with 4 (or however many GB of Ram you like, but this works well for me personally). Oh and remember to go into your directory wherever the GTFS & .pbf are in order to remove the other files that were generated--the .mapdb and mapdb.p files.

Basically, I think that error is because when you attempt to run the expanded time series, the program tries to call the network.dat file and if it can't find it, it breaks. So if you aren't able to generate that file during the setup, that's the first big problem to solve before moving on to the expanded time series. Hope that helps!

**Comment by rafapereirabr** [comment]
Hi @willj98 , 

1. could you please run `r5r::r5r_sitrep()` and share the output below?
2. could please try running the reprex below and let us know if it works?

```
library(r5r)

# build transport network
data_path <- system.file("extdata/poa", package = "r5r")
r5r_core <- setup_r5(data_path)

# load origin/destination points
points <- read.csv(file.path(data_path, "poa_points_of_interest.csv"))

departure_datetime <- as.POSIXct(
  "13-05-2019 14:00:00",
  format = "%d-%m-%Y %H:%M:%S"
)

ttm <- travel_time_matrix(
  r5r_core,
  origins = points,
  destinations = points,
  mode = c("WALK", "TRANSIT"),
  departure_datetime = departure_datetime,
  max_trip_duration = 60
)
head(ttm)
 
```

**Comment by rafapereirabr** [comment]
I would strongly suggest using [{rJavaEnv}](https://github.com/e-kotov/rJavaEnv) to install Java so it can be used with {r5r}.

In any case, I'm closing this issue due to no response. I'm happy to reopen it if the problem persists

**Comment by e-kotov** [comment]
> I would strongly suggest using [{rJavaEnv}](https://github.com/e-kotov/rJavaEnv) to install Java so it can be used with {r5r}.

Since our last exchange with @willj98 , I have fixed some bugs in rJavaEnv and improved the reliability. Perhaps it will work now for this specific case, if it had not before.
